### PR TITLE
Register block style early.

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/58528
---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/blocks/index.php [hunk 1]
```diff
@@ -19,6 +19,46 @@
...
+					$style_handle,
+					false
+				);
+				continue;
+			}
+			$style_handle = $schema[ $style_field ];
+			if ( is_array( $style_handle ) ) {
+				continue;
+			}
+			$path = "/wp-includes/blocks/$name/style$suffix.css";
```
Comment (westonruter) [inline src/wp-includes/blocks/index.php]:
```suggestion
			$path = "/wp-includes/blocks/$name/style{$suffix}.css";
```

File: src/wp-includes/blocks/index.php [hunk 2]
```diff
@@ -19,6 +19,46 @@
...
+
+	$style_fields = array(
+		'editorStyle',
+		'style',
+	);
+
+	foreach ( $core_blocks_meta as $name => $schema ) {
+		foreach ( $style_fields as $style_field ) {
+			if ( ! isset( $schema[ $style_field ] ) ) {
+				$style_handle = 'style' === $style_field ? "wp-block-$name" : "wp-block-{$name}-editor";
```
Comment (westonruter) [inline src/wp-includes/blocks/index.php]:
```suggestion
				$style_handle = 'style' === $style_field ? "wp-block-{$name}" : "wp-block-{$name}-editor";
```

**Review by westonruter — COMMENTED** [review]
Just a couple driveby nits

**Comment by spacedmonkey** [comment]
CC @aristath

File: src/wp-includes/blocks/index.php [hunk 3]
```diff
@@ -12,6 +12,54 @@
...
+	foreach ( $core_blocks_meta as $name => $schema ) {
+		foreach ( $style_fields as $style_field ) {
+			if ( ! isset( $schema[ $style_field ] ) ) {
+				$style_handle = 'style' === $style_field ? "wp-block-{$name}" : "wp-block-{$name}-editor";
+				wp_register_style(
+					$style_handle,
+					false
+				);
+				continue;
+			}
```
Comment (felixarntz) [inline src/wp-includes/blocks/index.php]: Why is this needed? In `register_block_style_handle()`, if the field doesn't exist in the block metadata, it just returns `false` and doesn't call `wp_register_style()`.

The logic used to potentially register with `false` URI is different in `register_block_style_handle()`, we should make sure to follow that here.

File: src/wp-includes/blocks/index.php [hunk 4]
```diff
@@ -12,6 +12,54 @@
...
+			$style_handle = $schema[ $style_field ];
+			if ( is_array( $style_handle ) ) {
+				continue;
+			}
+			$path = "/wp-includes/blocks/$name/style{$suffix}.css";
+			wp_register_style( $style_handle, $path );
```
Comment (felixarntz) [inline src/wp-includes/blocks/index.php]: This needs to use the URL like in `register_block_style_handle()` instead of path:
```suggestion
			wp_register_style( $style_handle, includes_url( "blocks/$name/style{$suffix}.css" );
```

File: src/wp-includes/blocks/index.php [hunk 5]
```diff
@@ -12,6 +12,54 @@
 require BLOCKS_PATH . 'widget-group.php';
 require BLOCKS_PATH . 'require-dynamic-blocks.php';
 
+/**
+ * Registers core block types styles.
+ *
+ * @since 6.3.0
+ */
+function register_core_block_types_styles() {
```
Comment (felixarntz) [inline src/wp-includes/blocks/index.php]: I think a better name would be `register_core_block_style_handles()`, because this function basically does what otherwise `register_block_style_handle()` would do, only in a more efficient way for core blocks.

File: src/wp-includes/blocks/index.php [hunk 6]
```diff
@@ -12,6 +12,54 @@
 require BLOCKS_PATH . 'widget-group.php';
 require BLOCKS_PATH . 'require-dynamic-blocks.php';
 
+/**
+ * Registers core block types styles.
```
Comment (felixarntz) [inline src/wp-includes/blocks/index.php]: It's worth clarifying why this function exists, compared to `register_block_style_handle()`.
```suggestion
 * Registers core block style handles.
 *
 * While {@see register_block_style_handle()} is typically used for that, the way it is
 * implemented is inefficient for core block styles. Registering those style handles here
 * avoids unnecessary logic and filesystem lookups in the other function.
```

**Review by felixarntz — CHANGES_REQUESTED** [review]
@spacedmonkey This looks great so far. Only a bit of feedback below.

File: src/wp-includes/blocks/index.php (hunk 4 above)
Comment (spacedmonkey) [reply src/wp-includes/blocks/index.php]: If you look at `wp_default_scripts`, other core styles do not do this. I am not sure why, but they do not. It is worth noting that most of these styles will never use this url for anything, as they will be inlines by `wp_maybe_inline_styles`. 

I do see why this should different from `wp_default_scripts`, but I could be talked out of it.

File: src/wp-includes/blocks/index.php (hunk 3 above)
Comment (spacedmonkey) [reply src/wp-includes/blocks/index.php]: This confused me as well, but consider. https://github.com/WordPress/wordpress-develop/commit/716a99587c04e5b37b4ceacc3ccce6c72c842464 / [https://core.trac.wordpress.org/ticket/53358 #53358] / https://github.com/WordPress/gutenberg/pull/32275/. 

Basically what happens in `register_block_style_handle` use realpath to check if the file exists. Look at this snippet 

https://github.com/WordPress/wordpress-develop/blob/fbef5030f67b6837119e9c890288bb0ace796940/src/wp-includes/blocks.php#L225-L255

If the file doesn't exist, then sets `$style_uri` to false. This is so that developers can add inline styles to these style sheets and style core blocks this way. Core itself also use these styles, see wp_common_block_scripts_and_styles. Example 

https://github.com/WordPress/wordpress-develop/blob/fbef5030f67b6837119e9c890288bb0ace796940/src/wp-includes/script-loader.php#L2371

With this, lots of themes / plugins and core itself would break. It is confusing, but it is needed. Maybe we should add a comment explaining it...

File: src/wp-includes/blocks/index.php [hunk 7]
```diff
@@ -12,6 +12,62 @@
...
+					$style_handle,
+					false
+				);
+				continue;
+			}
+			$style_handle = $schema[ $style_field ];
+			if ( is_array( $style_handle ) ) {
+				continue;
+			}
+			$path = "/wp-includes/blocks/$name/style{$suffix}.css";
```
Comment (mukeshpanchal27) [inline src/wp-includes/blocks/index.php]:
```suggestion
			$path = "/wp-includes/blocks/{$name}/style{$suffix}.css";
```

File: src/wp-includes/blocks/index.php [hunk 8]
```diff
@@ -12,6 +12,62 @@
...
+	if ( ! wp_should_load_separate_core_block_assets() ) {
+		return;
+	}
+
+	static $core_blocks_meta;
+	if ( ! $core_blocks_meta ) {
+		$core_blocks_meta = require ABSPATH . WPINC . '/blocks/blocks-json.php';
+	}
+
+	$suffix = SCRIPT_DEBUG ? '' : '.min';
```
Comment (mukeshpanchal27) [inline src/wp-includes/blocks/index.php]:
```suggestion
	$suffix = wp_scripts_get_suffix();
```
Can we use function instead of constant as core use function in many files.

File: src/wp-includes/blocks/index.php [hunk 9]
```diff
@@ -12,6 +12,62 @@
...
+			$wp_styles->add( $style_handle, $path );
+			$wp_styles->add_data( $style_handle, 'path', ABSPATH . $path );
+
+			$rtl_file = str_replace( "{$suffix}.css", "-rtl{$suffix}.css", ABSPATH . $path );
+			if ( is_rtl() && file_exists( $rtl_file ) ) {
+				$wp_styles->add_data( $style_handle, 'rtl', 'replace' );
+				$wp_styles->add_data( $style_handle, 'suffix', $suffix );
+				$wp_styles->add_data( $style_handle, 'path', $rtl_file );
+			}
+
```
Comment (mukeshpanchal27) [inline src/wp-includes/blocks/index.php]:
```suggestion
```

**Review by mukeshpanchal27 — COMMENTED** [review]
Great work @spacedmonkey, Left some feedback.

File: src/wp-includes/blocks.php [hunk 10]
```diff
@@ -205,6 +205,10 @@ function register_block_style_handle( $metadata, $field_name, $index = 0 ) {
 		$style_handle = $style_handle[ $index ];
 	}
 
+	if ( wp_style_is( $style_handle, 'registered' ) ) {
```
Comment (mukeshpanchal27) [inline src/wp-includes/blocks.php]: @spacedmonkey, could you please add a small description of this change so that others can understand it as well?

File: src/wp-includes/blocks/index.php [hunk 11]
```diff
@@ -12,6 +12,61 @@
...
+				);
+				continue;
+			}
+			$style_handle = $schema[ $style_field ];
+			if ( is_array( $style_handle ) ) {
+				continue;
+			}
+			$path = "/wp-includes/blocks/{$name}/style{$suffix}.css";
+			$wp_styles->add( $style_handle, $path );
+			$wp_styles->add_data( $style_handle, 'path', ABSPATH . $path );
```
Comment (mukeshpanchal27) [inline src/wp-includes/blocks/index.php]:
```suggestion
			$path = WPINC . "/blocks/{$name}/style{$suffix}.css";
			$wp_styles->add( $style_handle, "/{$path}" );
			$wp_styles->add_data( $style_handle, 'path', ABSPATH . $path );
```

`ABSPATH . $path` added duplicate forward slash. Let's follow similar approach that used here https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/script-loader.php#L1600-L1602

**Image:**
Screenshot 2023-06-20 at 10 18 50 AM

File: src/wp-includes/blocks/index.php [hunk 12]
```diff
@@ -12,6 +12,61 @@
...
+
+			$rtl_file = str_replace( "{$suffix}.css", "-rtl{$suffix}.css", ABSPATH . $path );
```
Comment (mukeshpanchal27) [inline src/wp-includes/blocks/index.php]: Could we move `str_replace` code inside the `is_rtl()` condition?  this way we reduce 184 replacement of style for normal sites and only perform it for RTL sites.

**Review by mukeshpanchal27 — COMMENTED** [review]
@spacedmonkey A few last feedback, but almost good to go.

File: src/wp-includes/blocks/index.php (hunk 12 above)
Comment (spacedmonkey) [reply src/wp-includes/blocks/index.php]: This code is copied from elsewhere. I think we should leave as is. 

https://github.com/WordPress/wordpress-develop/blob/76376fdbc3dc0b3261de377dffc350677345e7ba/src/wp-includes/blocks.php#L263-L273

File: src/wp-includes/blocks/index.php [hunk 13]
```diff
@@ -12,6 +12,63 @@
...
+				);
+				continue;
+			}
+			$style_handle = $schema[ $style_field ];
+			if ( is_array( $style_handle ) ) {
+				continue;
+			}
+			$style_path = "blocks/{$name}/style{$suffix}.css";
+			$wp_styles->add( $style_handle, $includes_url . $style_path );
+			$path = ABSPATH . WPINC . DIRECTORY_SEPARATOR . $style_path;
```
Comment (mukeshpanchal27) [inline src/wp-includes/blocks/index.php]: @spacedmonkey Core avoid to use `DIRECTORY_SEPARATOR` constant. See https://core.trac.wordpress.org/ticket/44654

File: src/wp-includes/blocks/index.php [hunk 14]
```diff
@@ -12,6 +12,63 @@
...
+	if ( ! wp_should_load_separate_core_block_assets() ) {
+		return;
+	}
+
+	static $core_blocks_meta;
+	if ( ! $core_blocks_meta ) {
+		$core_blocks_meta = require ABSPATH . WPINC . '/blocks/blocks-json.php';
+	}
+
+	$includes_url = includes_url();
```
Comment (mukeshpanchal27) [inline src/wp-includes/blocks/index.php]:
```suggestion
	$includes_url = includes_url( '', 'relative' );
```
Not blocker. before this changes it was `/wp-includes/blocks/blockbame/style.css, if we use `includes_url()` it return include URL link.

**Review by mukeshpanchal27 — COMMENTED** [review]
@spacedmonkey Left some feedback.

File: src/wp-includes/blocks/index.php (hunk 13 above)
Comment (spacedmonkey) [inline src/wp-includes/blocks/index.php]:
```suggestion
			$path = ABSPATH . WPINC . '/' . $style_path;
```

File: src/wp-includes/blocks/index.php (hunk 14 above)
Comment (spacedmonkey) [reply src/wp-includes/blocks/index.php]: This was added becuase of feedback from @felixarntz https://github.com/WordPress/wordpress-develop/pull/4621#discussion_r1232574807

I don't see how it hurts to include the url here.

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks @spacedmonkey for the update. It look good to me.

**Comment by mukeshpanchal27** [comment]
@spacedmonkey Re-run the failed job it show `Navigation timeout of 30000 ms exceeded`.

File: src/wp-includes/blocks/index.php (hunk 4 above)
Comment (felixarntz) [reply src/wp-includes/blocks/index.php]: Better to keep this like it was before / like it is in `register_block_style_handle()`. I see you updated it, so all good now 👍

File: src/wp-includes/blocks/index.php (hunk 3 above)
Comment (felixarntz) [reply src/wp-includes/blocks/index.php]: @spacedmonkey I understand, however I am referring more specifically to the exact logic: There are three possible scenarios technically:
1. The `style`/`editorStyle` properties are defined in the schema _and_ the files exist.
2. The `style`/`editorStyle` properties are defined in the schema, but the files do not exist.
3. The `style`/`editorStyle` properties are not defined in the schema.

Right now, the PR registers the stylesheet with `false` in case 3 above. However, `register_block_style_handle()` does nothing in case 3, it bails early. Instead, it registers the stylesheet with `false` in case 2 above. So that is where your implementation and the current one in core differ.

I am not entirely sure how relevant that is for core blocks. Can you do some testing with your PR and check which core blocks (if any) the condition in https://github.com/WordPress/wordpress-develop/pull/4621/files#diff-c62beff71446448ff952e38805c8ae7652f4e2717d3ca41b1d9cd9963bd82464R44 applies to? That would help clarify whether this code is okay as is or whether it's safer to remove it.

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey Looks good except for the pending conversation where I just left a follow-up on https://github.com/WordPress/wordpress-develop/pull/4621#discussion_r1237444923. Maybe it's okay as is, but we have to validate that, since right now your logic is a bit different from the existing one. So we have to make sure we don't accidentally introduce a regression there.

File: src/wp-includes/blocks/index.php [hunk 15]
```diff
@@ -12,6 +12,77 @@
...
+	$style_fields   = array(
+		'style',
+		'editorStyle',
+	);
+	$transient_name = 'block_styles';
+	$files          = get_transient( $transient_name );
+	if ( ! $files ) {
+		$files = glob( __DIR__ . "/**/style*{$suffix}.css" );
+		set_transient( $transient_name, $files );
+	}
```
Comment (spacedmonkey) [inline src/wp-includes/blocks/index.php]: This checks to see if the file exists once and then caches.

File: src/wp-includes/blocks/index.php [hunk 16]
```diff
@@ -12,6 +12,77 @@
...
+			$schema['editorStyle'] = "wp-block-{$name}-editor";
+		}
+		foreach ( $style_fields as $style_field ) {
+			$style_handle = $schema[ $style_field ];
+			if ( is_array( $style_handle ) ) {
+				continue;
+			}
+
+			$style_path = "blocks/{$name}/style{$suffix}.css";
+			$path       = ABSPATH . WPINC . '/' . $style_path;
```
Comment (felixarntz) [inline src/wp-includes/blocks/index.php]: Nit-pick: Maybe defined a `$includes_path = ABSPATH . WPINC . '/'` variable outside of the loop to improve readability. Not critical, but you already have `$includes_url`, so this would be nice for parity.

File: src/wp-includes/blocks/index.php [hunk 17]
```diff
@@ -12,6 +12,77 @@
...
+	}
+
+	$includes_url   = includes_url();
+	$suffix         = wp_scripts_get_suffix();
+	$wp_styles      = wp_styles();
+	$style_fields   = array(
+		'style',
+		'editorStyle',
+	);
+	$transient_name = 'block_styles';
```
Comment (felixarntz) [inline src/wp-includes/blocks/index.php]: Since the styles are all from core itself, I think we can easily handle invalidation here when WordPress is updated: Either we add `$wp_version` to the transient name, or we delete the transient on every WordPress update (like you're doing above). In order to support programmatically controlled sites (e.g. that update via Composer), it may be more reliable to rely on the `$wp_version` global.

Also, let's prefix this with something like `wp_core_`, to clarify that it is only for WP core block styles.
```suggestion
	$transient_name = 'wp_core_block_styles_' . $GLOBALS['wp_version'];
```

File: src/wp-includes/blocks/index.php [hunk 18]
```diff
@@ -12,6 +12,77 @@
...
+		set_transient( $transient_name, $files );
+	}
+
+	foreach ( $core_blocks_meta as $name => $schema ) {
+		if ( ! isset( $schema['style'] ) ) {
+			$schema['style'] = "wp-block-$name";
+		}
+		if ( ! isset( $schema['editorStyle'] ) ) {
+			$schema['editorStyle'] = "wp-block-{$name}-editor";
+		}
```
Comment (felixarntz) [inline src/wp-includes/blocks/index.php]: Before this backfilling, we also need to run the `block_type_metadata` filter, otherwise the data may become inconsistent.
```suggestion
		/** This filter is documented in wp-includes/blocks.php */
		$schema = apply_filters( 'block_type_metadata', $schema );

		// Backfill these properties similar to `register_block_type_from_metadata()`.
		if ( ! isset( $schema['style'] ) ) {
			$schema['style'] = "wp-block-$name";
		}
		if ( ! isset( $schema['editorStyle'] ) ) {
			$schema['editorStyle'] = "wp-block-{$name}-editor";
		}
```

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey This is pretty much good to go, after another review the changes make sense for parity with the existing `register_block_style_handle()`. A few outstanding points of feedback below.

Note that I found a bug that is related to this code, see https://core.trac.wordpress.org/ticket/58605.

File: src/wp-includes/blocks/index.php (hunk 17 above)
Comment (spacedmonkey) [reply src/wp-includes/blocks/index.php]: @felixarntz I originally had it like this. But as the tranient would be stored in options. This would mean that a version of the transient would be stored in an autoloaded option forever more right?

Comment (felixarntz) [reply src/wp-includes/blocks/index.php]: @spacedmonkey Fair point, in that case, let's just change to 
```suggestion
	$transient_name = 'wp_core_block_styles';
```
and use invalidation on update.

**Comment by spacedmonkey** [comment]
@felixarntz @mukeshpanchal27 I went down a habbit hole. I fixed two more issues. First there is a fix for [https://core.trac.wordpress.org/ticket/58605](https://core.trac.wordpress.org/ticket/58605). Changing the glob, fixes this. 
Secondly, looking at the function `wp_common_block_scripts_and_styles`. This too does a glob on css files as well. It doesn't make sense to do a glob in two places. So I moved this logic to this function as well. This has some really nice saves as well. To test this, you should switch to the tt2 theme. 

This PR now needs some unit tests, I will work on the ASAP.

**Comment by felixarntz** [comment]
@spacedmonkey I don't think it's a good idea to intertwine the two tickets in this pull request. Yes, they are related but we're also close to the 6.3 beta, and you now added lots of new logic that changed the PR notably from what it was before.

I would have approved it now in its original scope, but the new changes make this a whole lot more complicated and potentially risky since you moved logic from one place to the other. I would suggest that you revert those changes and open a separate PR to address https://core.trac.wordpress.org/ticket/58560. That ticket is much less critical, and it's not even in the 6.3 milestone. Let's tackle it separately. Then we can get this PR merged in its original scope.

File: src/wp-includes/blocks/index.php [hunk 19]
```diff
@@ -12,6 +12,82 @@
...
+	$wp_styles      = wp_styles();
+	$style_fields   = array(
+		'style'       => 'style',
+		'editorStyle' => 'editor',
+	);
+	$transient_name = 'wp_core_block_styles';
+	$files          = get_transient( $transient_name );
+	if ( ! $files ) {
+		$files = glob( __DIR__ . "/**/**{$suffix}.css" );
+		set_transient( $transient_name, $files );
```
Comment (spacedmonkey) [inline src/wp-includes/blocks/index.php]: @felixarntz Should we disable this cache is the WP_DEBUG is enabled?

**Comment by spacedmonkey** [comment]
@felixarntz Fair point, adding back is simple enough. I will do it as part of that release. I have reverted the changes now, can you take a look again.

File: src/wp-includes/blocks/index.php (hunk 19 above)
Comment (felixarntz) [reply src/wp-includes/blocks/index.php]: @spacedmonkey Sounds good. Please add the same `todo` comment that is present on all the other places where we do caching around e.g. the `theme.json` stuff.

Related: https://core.trac.wordpress.org/ticket/57487

File: src/wp-includes/blocks/index.php [hunk 20]
```diff
@@ -12,6 +12,82 @@
...
+				$wp_styles->add( $style_handle, $includes_url . $style_path );
+				$wp_styles->add_data( $style_handle, 'path', $path );
+
+				$rtl_file = str_replace( "{$suffix}.css", "-rtl{$suffix}.css", $path );
+				if ( is_rtl() && in_array( $rtl_file, $files, true ) ) {
+					$wp_styles->add_data( $style_handle, 'rtl', 'replace' );
+					$wp_styles->add_data( $style_handle, 'suffix', $suffix );
+					$wp_styles->add_data( $style_handle, 'path', $rtl_file );
+				}
+			}
```
Comment (felixarntz) [inline src/wp-includes/blocks/index.php]: Using if-else is not a great pattern, and we could easily avoid it here. I think you previously had it like that, not sure why it changed. Better for readability to use early returns than nesting.
```suggestion
			if ( ! in_array( $path, $files, true ) ) {
				$wp_styles->add(
					$style_handle,
					false
				);
				continue;
			}

			$wp_styles->add( $style_handle, $includes_url . $style_path );
			$wp_styles->add_data( $style_handle, 'path', $path );

			$rtl_file = str_replace( "{$suffix}.css", "-rtl{$suffix}.css", $path );
			if ( is_rtl() && in_array( $rtl_file, $files, true ) ) {
				$wp_styles->add_data( $style_handle, 'rtl', 'replace' );
				$wp_styles->add_data( $style_handle, 'suffix', $suffix );
				$wp_styles->add_data( $style_handle, 'path', $rtl_file );
			}
```

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey Thanks for reverting, this is pretty much good to go. Only outstanding bit is https://github.com/WordPress/wordpress-develop/pull/4621/files#r1240041002, plus I have one nit-pick below - not a blocker, but it would be good to change.

File: src/wp-includes/blocks/index.php (hunk 19 above)
Comment (spacedmonkey) [reply src/wp-includes/blocks/index.php]: @felixarntz Pushed the change.

File: src/wp-includes/blocks/index.php (hunk 20 above)
Comment (spacedmonkey) [reply src/wp-includes/blocks/index.php]: Fixed.

File: src/wp-includes/blocks/index.php [hunk 21]
```diff
@@ -12,6 +12,93 @@
...
+	 * @todo Replace `WP_DEBUG` once an "in development mode" check is available in Core.
+	 */
+	if ( ! WP_DEBUG ) {
+		$transient_name = 'wp_core_block_styles';
+		$files          = get_transient( $transient_name );
+		if ( ! $files ) {
+			$files = glob( __DIR__ . "/**/**.css" );
+			set_transient( $transient_name, $files );
+		}
+	}
```
Comment (felixarntz) [inline src/wp-includes/blocks/index.php]: I think you have a bug here now, if `WP_DEBUG` is enabled, `$files` will always be an empty array.
```suggestion

	/*
	 * Ignore transient cache when `WP_DEBUG` is enabled. Why? To avoid interfering with
	 * the core developer's workflow.
	 *
	 * @todo Replace `WP_DEBUG` once an "in development mode" check is available in Core.
	 */
	if ( ! WP_DEBUG ) {
		$transient_name = 'wp_core_block_styles';
		$files          = get_transient( $transient_name );
		if ( ! $files ) {
			$files = glob( __DIR__ . "/**/**.css" );
			set_transient( $transient_name, $files );
		}
	} else {
		$files = glob( __DIR__ . "/**/**.css" );
	}
```

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey One last thing, I think there's a bug in the new code you added.

**Comment by felixarntz** [comment]
@spacedmonkey FYI I just committed https://core.trac.wordpress.org/changeset/56005, which you have already accounted for here in https://github.com/WordPress/wordpress-develop/pull/4621/files#diff-c62beff71446448ff952e38805c8ae7652f4e2717d3ca41b1d9cd9963bd82464R38-R41 👍 

Maybe refresh your PR branch against `trunk` to make sure tests still pass.

File: src/wp-includes/blocks.php [hunk 22]
```diff
@@ -368,6 +373,7 @@ function register_block_type_from_metadata( $file_or_folder, $args = array() ) {
 		if ( ! isset( $metadata['style'] ) ) {
 			$metadata['style'] = "wp-block-$block_name";
 		}
+
```
Comment (spacedmonkey) [inline src/wp-includes/blocks.php]:
```suggestion
```

**Review by felixarntz — APPROVED** [review]
Thanks @spacedmonkey, LGTM in terms of production code!

If you can add a few tests, that would be excellent 🎉

File: src/wp-includes/blocks/index.php [hunk 23]
```diff
@@ -12,6 +12,94 @@
...
+		$files = glob( __DIR__ . '/**/**.css' );
+	}
+
+	foreach ( $core_blocks_meta as $name => $schema ) {
+		/** This filter is documented in wp-includes/blocks.php */
+		$schema = apply_filters( 'block_type_metadata', $schema );
+
+		// Backfill these properties similar to `register_block_type_from_metadata()`.
+		if ( ! isset( $schema['style'] ) ) {
+			$schema['style'] = "wp-block-$name";
```
Comment (mukeshpanchal27) [inline src/wp-includes/blocks/index.php]:
```suggestion
			$schema['style'] = "wp-block-{$name}";
```

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks @spacedmonkey, Great work. Approved with one nit-pick.

File: src/wp-includes/blocks/index.php (hunk 23 above)
Comment (spacedmonkey) [inline src/wp-includes/blocks/index.php]:
```suggestion
			$schema['style'] = "wp-block-{$name}";
```

File: src/wp-includes/blocks.php (hunk 22 above)
Comment (spacedmonkey) [inline src/wp-includes/blocks.php]:
```suggestion
```

**Comment by spacedmonkey** [comment]
@felixarntz @mukeshpanchal27 As requested, I have added unit tests. The unit tests have highlighted that the incorrectly value was being return. I moved the some the logic in `register_block_style_handle` around. This does two things, one skips the need to call `wp_should_load_separate_core_block_assets`, which resulted in a 1ms of load time. Two, calls `generate_block_asset_handle` and gets the correct style handle name early, so it can be used with `wp_style_is`. 

This change now is even better of a performance improvement.

File: tests/phpunit/tests/blocks/registerCoreBlockStyleHandles.php [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+		register_core_block_style_handles();
+
+		$style_fields = array(
+			'style'       => 'style',
+			'editorStyle' => 'editor',
+		);
+
+		$core_blocks_meta = $this->get_block_data();
+
+		foreach ( $core_blocks_meta as $name => $schema ) {
```
Comment (felixarntz) [inline tests/phpunit/tests/blocks/registerCoreBlockStyleHandles.php]: Instead of a `foreach`, this test should use a data provider. You can change the `get_block_data()` method to become such a data provider, e.g. `data_get_core_blocks()`, and then it returns an array, which for every block includes the `$name` and `$schema`.

File: tests/phpunit/tests/blocks/registerCoreBlockStyleHandles.php [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+		$style_fields = array(
+			'style'       => 'style',
+			'editorStyle' => 'editor',
+		);
+
+		$core_blocks_meta = $this->get_block_data();
+		$wp_styles        = $GLOBALS['wp_styles'];
+		$includes_url     = includes_url();
+
+		foreach ( $core_blocks_meta as $name => $schema ) {
```
Comment (felixarntz) [inline tests/phpunit/tests/blocks/registerCoreBlockStyleHandles.php]: Same as above, this should use a data provider method instead of having lots of assertions in a single test run going through a loop. You can use the same data provider method for this and the above test method.

File: tests/phpunit/tests/blocks/registerCoreBlockStyleHandles.php [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+
+			if ( ! isset( $schema['style'] ) ) {
+				$schema['style'] = "wp-block-$name";
+			}
+			if ( ! isset( $schema['editorStyle'] ) ) {
+				$schema['editorStyle'] = "wp-block-{$name}-editor";
+			}
```
Comment (felixarntz) [inline tests/phpunit/tests/blocks/registerCoreBlockStyleHandles.php]: Another benefit of using a data provider is that you'll be able to do this in the data provider method rather than duplicated across both tests.

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey Production code LGTM! For the tests, please use a data provider rather than a big `foreach` loop in the tests themselves.

File: src/wp-includes/blocks/index.php [hunk 27]
```diff
@@ -12,6 +12,94 @@
...
+	);
+
+	/*
+	 * Ignore transient cache when `WP_DEBUG` is enabled. Why? To avoid interfering with
+	 * the core developer's workflow.
+	 *
+	 * @todo Replace `WP_DEBUG` once an "in development mode" check is available in Core.
+	 */
+	if ( ! WP_DEBUG ) {
+		$transient_name = 'wp_core_block_styles';
```
Comment (westonruter) [inline src/wp-includes/blocks/index.php]: Would this be better named:

```suggestion
		$transient_name = 'wp_core_block_css_files';
```

File: src/wp-admin/includes/upgrade.php [hunk 28]
```diff
@@ -654,6 +654,8 @@ function wp_upgrade() {
 			update_site_meta( get_current_blog_id(), 'db_last_updated', microtime() );
 		}
 
+		delete_transient( 'wp_core_block_styles' );
```
Comment (westonruter) [inline src/wp-admin/includes/upgrade.php]:
```suggestion
		delete_transient( 'wp_core_block_css_files' );
```

File: src/wp-includes/blocks/index.php [hunk 29]
```diff
@@ -12,6 +12,94 @@
...
+		$files          = get_transient( $transient_name );
+		if ( ! $files ) {
+			$files = glob( __DIR__ . '/**/**.css' );
+			set_transient( $transient_name, $files );
```
Comment (westonruter) [inline src/wp-includes/blocks/index.php:52]: Seems like this could be setting a site transient instead:

```suggestion
			set_site_transient( $transient_name, $files );
```

File: src/wp-admin/includes/upgrade.php (hunk 28 above)
Comment (westonruter) [reply src/wp-admin/includes/upgrade.php]: Also, per below, wouldn't site transients be more appropriate?

```suggestion
		delete_site_transient( 'wp_core_block_css_files' );
```

Comment (spacedmonkey) [reply src/wp-admin/includes/upgrade.php]: @westonruter Using network options ( tranisents ) has a performance impact. I think network options are not autoloaded. So it result in a database query, slow things down. Maybe we could work this out later.

File: src/wp-includes/blocks/index.php (hunk 29 above)
Comment (spacedmonkey) [reply src/wp-includes/blocks/index.php:52]: See https://github.com/WordPress/wordpress-develop/pull/4621/files#r1242531400

**Review by felixarntz — APPROVED** [review]
Thanks @spacedmonkey, looks great! Good to commit IMO.

File: src/wp-includes/blocks/index.php [hunk 30]
```diff
@@ -12,6 +12,94 @@
...
+		'editorStyle' => 'editor',
+	);
+
+	/*
+	 * Ignore transient cache when `WP_DEBUG` is enabled. Why? To avoid interfering with
+	 * the core developer's workflow.
+	 *
+	 * @todo Replace `WP_DEBUG` once an "in development mode" check is available in Core.
+	 */
+	if ( ! WP_DEBUG ) {
```
Comment (felixarntz) [inline src/wp-includes/blocks/index.php]: Now that https://github.com/WordPress/wordpress-develop/commit/4a16702090984caef24af5a07d598cc5afff2fdc has been committed, can you please refresh this PR against latest `trunk` and replace this?
```suggestion
	/*
	 * Ignore transient cache when the development mode is set to 'core', so it doesn't interfere with the core
	 * developer's workflow.
	 */
	if ( wp_get_development_mode() !== 'core' ) {
```

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey One small tweak here is needed now that `wp_get_development_mode()` is available in `trunk`, see below. Can you please update the code as suggested to get rid of the todo comment?

It would be nice to add a simple test to ensure that the transient is used only when the development mode is not "core", so if you can implement that before the beta release, that would be awesome, though not a blocker IMO. For reference, it could be similar to e.g. https://github.com/WordPress/wordpress-develop/commit/4a16702090984caef24af5a07d598cc5afff2fdc#diff-a2793e725e03902d8a665c105916351677858253a237e4d08bdc1c01f8c3d671R38-R61.

**Comment by spacedmonkey** [comment]
Committed https://github.com/WordPress/wordpress-develop/commit/d8409a205a40a9c5a69d668a9aa2aa3123f4c3e5
